frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -144,49 +144,63 @@ impl<T: Config> Pallet<T> {
 			Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
 		let desired_targets = Self::desired_targets().ok_or(MinerError::SnapshotUnAvailable)?;
 
-		// closures.
+		// now make some helper closures.
 		let cache = helpers::generate_voter_cache::<T>(&voters);
 		let voter_index = helpers::voter_index_fn::<T>(&cache);
 		let target_index = helpers::target_index_fn::<T>(&targets);
 		let voter_at = helpers::voter_at_fn::<T>(&voters);
 		let target_at = helpers::target_at_fn::<T>(&targets);
 		let stake_of = helpers::stake_of_fn::<T>(&voters, &cache);
 
-		let ElectionResult { assignments, winners } = election_result;
+		// Compute the size of a compact solution comprised of the selected arguments.
+		//
+		// This function completes in `O(edges)`; it's expensive, but linear.
+		let encoded_size_of = |assignments: &[IndexAssignmentOf<T>]| {
+			CompactOf::<T>::try_from(assignments).map(|compact| compact.encoded_size())
+		};
+
+		let ElectionResult { mut assignments, winners } = election_result;
+		// Sort the assignments by reversed voter stake. This ensures that we can efficiently truncate the list.
+		let stakes: BTreeMap<_, _> =
```

Comment:
mm seems duplicate to what we generate in `cache`, so here we have `BtreeMap<Acc, Stake>`, and previously we already generated `BtreeMap<Acc, index>` and then can use the index to directly go to `snapshots`. Haven't really thought about complexity, but the allocation can def. be avoided.

Reply:
I see what you mean, yeah. dd80c95be. Should have the same complexity as a direct lookup, and avoid the allocation.

Reply:
sweet.